src/cuda.rs

```diff
@@ -0,0 +1,98 @@
+pub type c_char = u8;
+pub type size_t = usize;
+
+cfg_if! {
+    if #[cfg(libc_core_cvoid)] {
+        pub use ::ffi::c_void;
+    } else {
+        // Use repr(u8) as LLVM expects `void*` to be the same as `i8*` to help
+        // enable more optimization opportunities around it recognizing things
+        // like malloc/free.
+        #[repr(u8)]
+        #[allow(missing_copy_implementations)]
+        #[allow(missing_debug_implementations)]
+        pub enum c_void {
+            // Two dummy variants so the #[repr] attribute can be used.
+            #[doc(hidden)]
+            __variant1,
+            #[doc(hidden)]
+            __variant2,
+        }
+    }
+}
+
+extern "C" {
+    /// Syscall to be used whenever the *assert expression produces a `false` value*.
+    ///
+    /// Syscall arguments:
+    /// * `message`: The pointer to the string that should be output.
+    /// * `file`: The pointer to the file name string associated with the assert.
+    /// * `line`: The line number associated with the assert.
+    /// * `function`: The pointer to the function name string associated with the assert.
+    /// * `char_size`: The size in bytes of the characters contained in the provided strings. The only supported character size is 1.
+    ///
+    /// Source:
+    /// [PTX Interoperability](https://docs.nvidia.com/cuda/ptx-writers-guide-to-interoperability/index.html#system-calls).
+    pub fn __assertfail(
+        message: *const c_char,
+        file: *const c_char,
+        line: i32,
+        function: *const c_char,
+        char_size: size_t,
+    );
+
+    /// Print formatted output from a kernel to a host-side output stream.
+    ///
+    /// Syscall arguments:
+    /// * `status`: The status value that is returned by `vprintf`.
+    /// * `format`: A pointer to the format specifier input (uses common `printf` format).
+    /// * `valist`: A pointer to the valist input.
+    ///
+    /// ```
+    /// #[repr(C)]
+    /// struct PrintArgs(f32, f32, f32, i32);
+    ///
+    /// vprintf(
+    ///     "int(%f + %f) = int(%f) = %d\n".as_ptr(),
+    ///     transmute(&PrintArgs(a, b, a + b, (a + b) as i32)),
+    /// );
+    /// ```
+    ///
+    /// Sources:
+    /// [Programming Guide](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#formatted-output),
+    /// [PTX Interoperability](https://docs.nvidia.com/cuda/ptx-writers-guide-to-interoperability/index.html#system-calls).
+    pub fn vprintf(format: *const c_char, valist: *const c_void) -> i32;
```

Comment:
In clang `valist` is also a `*const c_char` and the return type is a `c_int`.

Reply:
Is there any difference from FFI or safety point of view between `*const c_char` and `*const c_void`?

From the logic point of view, having `*const c_void` is more correct because the second argument can be of any type, e.g. simple case:
``` rust
vprintf("c_int: %d".as_ptr(), transmute(&10));
```

Reply:
So I don't know actually. I kind of wish NVIDIA had a CUDA header for their "kernel-only" APIs that we could just follow here.

Reply:
I only found a meention of `printf` declaration in original CUDA headers:
``` c
__host__ __device__ __device_builtin__ __cudart_builtin__ int __cdecl printf(const char*, ...);
```

At least for now it seems to be safe to use `*const c_void` because LLVM emits same PTX with both types:
```
.extern .func  (.param .b32 func_retval0) vprintf
(
	.param .b64 vprintf_param_0,
	.param .b64 vprintf_param_1
)
```